accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+        // Create the directory and restrict access using Windows
+        // Access Control Lists (ACLs).
+
+        var rules = new DirectorySecurity();
+        rules.AddAccessRule(
+            new FileSystemAccessRule(@"fabrikam\log-readers",
+                                     FileSystemRights.Read,
+                                     AccessControlType.Allow)
+        );
+        rules.AddAccessRule(
+            new FileSystemAccessRule(@"fabrikam\log-writers",
+                                     FileSystemRights.FullControl,
+                                     AccessControlType.Allow)
+        );
+
+        Directory.CreateDirectory(loggingDirectory, rules);
+    }
+
+    return loggingDirectory;
+}
+```
+
+After that, the diagnostic disappears automatically.
+
+## Requirements
+
+### Goals
+
+* Developers get diagnostics when they inadvertently use platform-specific APIs.
+    - All diagnostics are provided on the command line as well as in the IDE
+    - The developer doesn't have to resort to explicit suppressions to indicate
+      intent. Idiomatic gestures, such as guarding the call with a platform
+      check or marking a method as platform-specific, automatically suppress
+      these diagnostics. However, explicit suppressions via the usual means
+      (`NoWarn`, `#pragma`) will also work.
+* The analyzer helps library authors to express their intent so that their
+  consumers can also benefit from these diagnostics.
+    - Library authors can annotate their platform-specific APIs without taking a
+      dependency.
+    - Library consumers don't have to install a NuGet package to get diagnostics
+      when using platform-specific APIs
+
+### Non-Goals
+
+* Modelling partially portable APIs (that is APIs are applicable to more than
+  one platform but no all platforms). Those should Be modeled as capability
+  APIs, as as `RuntimeInformation.IsDynamicCodeSupported` for ref emit. This
+  would be a separate analyzer.
+* Shipping the platform-specific analyzer and/or annotations out-of-band.
+
+## Design
+
+### Attribute
+
+OS bindings and platforms-specific BCL APIs will be annotated using a set of
+custom attributes:
+
+```C#
+namespace System.Runtime.InteropServices
+{
+    public partial struct OSPlatform
+    {
+        // Existing properties
+        // public static OSPlatform FreeBSD { get; }
+        // public static OSPlatform Linux { get; }
+        // public static OSPlatform OSX { get; }
+        // public static OSPlatform Windows { get; }
+        public static OSPlatform Android { get; }
+        public static OSPlatform iOS { get; }
+    }
+
+    public partial static class RuntimeInformation
+    {
+        // Existing API
+        // public static bool IsOSPlatform(OSPlatform osPlatform);
+
+        // Check for the OS with a >= version comparison
+        // Used to guard APIs that were added in the given OS release.
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major);
```

Comment:
What set of version numbers are you going to use for which numbers `major` and `minor` map to? Are you going to use the "marking" version numbers, or what the system returns as version numbers?

For example - on my macOS Mojave computer, if I print out `System.Environment.OSVersion.Version`, it says `18.7.0.0`. But the "About This Mac" window says `10.14.6`.

Similarly, on `Windows 8.1` if I print `System.Environment.OSVersion.Version`, it prints `6.3`. So I'm assuming these version numbers will use "Marketing" version numbers, since that is what a normal developer would expect.

What version numbers will `OSPlatform.Linux` compare against? The Kernel version?

Reply:
Since we fixed `Environment.OSVersion.Version` on Windows 10 with https://github.com/dotnet/runtime/issues/33643, I opened https://github.com/dotnet/runtime/issues/34977 to fix it on macOS as well. I assume on iOS, we will want `Environment.OSVersion.Version` to return numbers like `13.3.1` (iOS version) and not `19.3` (Darwin version)

Reply:
> What set of version numbers are you going to use for which numbers `major` and `minor` map to?

This is up to the SDK author. I don't know what Apple uses in their official documentation, but whatever version numbers the iOS SDK are the ones we should be using in the reference assembly for the iOS bindings. Since Apple expects developers to do version checks *somehow* they must have an API that does this kind of checking.

Reply:
I have added question in the Q&A section that answers this.

Reply:
> This is up to the SDK author.

It isn't really up to the SDK author - it is up to the runtime. `RuntimeInformation` is part of the runtime libraries, and this method needs to return a boolean. So `RuntimeInformation` itself is going to have to answer the question "is the current process running on `OSPlatform.Windows, 10, 0, 18363` or later"? Thus, it is `RuntimeInformation` that is going to have to decide which version numbers are appropriate to pass into this method.

If you want the SDK author to be able to define this, `RuntimeInformation` will have to either call out to the SDK author, or the SDK author will have to provide the current information to `RuntimeInformation`.

Reply:
It's the provider of the platform, be that Apple (iOS), Microsoft (Windows), Google (Android). They decide what version the developer has to check for (marketing vs. something else). That also means they will a way to check for this version at runtime.

Reply:
But this isn't an iOS, Windows, or Android API - it is a .NET runtime API. We are the ones that need to decide how to map this API to the underlying platform's API(s). For example, on Windows are we going to use [GetVersionEx](https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversionexw) or [RtlGetVersion](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/wdm/nf-wdm-rtlgetversion)? This is something the designer of this .NET runtime API needs to decide.

Maybe an alternative design would be to have platform specific APIs to do these checks, instead of forcing all platforms to use the same API. Then iOS apps can do the checks like iOS expects. Android apps can do checks like Android expects. And Windows apps can do the checks like Windows expects.
This could also address @timheuer's comment: https://github.com/dotnet/designs/pull/110#issuecomment-613521391, since Microsoft (Windows) doesn't suggest to check version numbers, but instead [check for capabilities](https://docs.microsoft.com/en-us/windows/uwp/debug-test-perf/version-adaptive-code).